tempesta_fw/sock_srv.c

```diff
@@ -181,6 +181,8 @@ tfw_sock_srv_connect_try(TfwSrvConn *srv_conn)
 		return r == SS_SHUTDOWN ? 0 : r;
 	}
 
+	/* Link Tempesta with the socket. */
+	tfw_connection_link_to_sk((TfwConn *)srv_conn, sk);
```

Comment:
Also, it's important here that when Tempesta is stopped, `ss_close_sync(conn->sk, true);` always pushes the request for a close to the work queue, so the closing in never synchronous. That's because Tempesta is stopped while in user context, and the condition `!in_softirq()` makes `__ss_close()` do just that. That way the case of `TCP_SYN_SENT` needs to be handled only in `ss_tx_action()`.

Perhaps, a note in the description of `tfw_sock_srv_disconnect()` would help to understand the intricacies of the behaviour better.